src/typecheck/mod.rs

```diff
@@ -1314,6 +1314,52 @@ fn row_add(
     }
 }
 
+/// Get the set of all unification variables in a type.
+fn get_free_variables(wildcards: &Vec<TypeWrapper>, t: TypeWrapper) -> HashSet<usize> {
+    match t {
+        TypeWrapper::Ptr(i) => HashSet::from([i]),
+        TypeWrapper::Concrete(ty) => match ty {
+            AbsType::Dyn()
+            | AbsType::Num()
+            | AbsType::Bool()
+            | AbsType::Str()
+            | AbsType::Sym()
+            | AbsType::Flat(_)
+            | AbsType::Var(_)
+            | AbsType::RowEmpty() => HashSet::new(),
+            // TODO: replace with get_wilccard_var(..)
+            AbsType::Wildcard(id) => wildcards
+                .get(id)
+                .map(|ty| get_free_variables(wildcards, ty.clone()))
+                .unwrap_or(HashSet::new()),
+            AbsType::Arrow(ty1, ty2) => {
+                let a = get_free_variables(wildcards, *ty1);
+                let b = get_free_variables(wildcards, *ty2);
+                a.union(&b).map(|i| *i).collect()
+            }
+            AbsType::Forall(_, ty)
+            | AbsType::Enum(ty)
+            | AbsType::StaticRecord(ty)
+            | AbsType::DynRecord(ty)
+            | AbsType::Array(ty) => get_free_variables(wildcards, *ty),
+            AbsType::RowExtend(_, ty_row, tail) => ty_row
+                .map(|ty| get_free_variables(wildcards, *ty))
+                .unwrap_or(HashSet::new())
+                .union(&get_free_variables(wildcards, *tail))
+                .map(|i| *i)
+                .collect(),
+        },
+        TypeWrapper::Contract(..) | TypeWrapper::Constant(_) => HashSet::new(),
+    }
+}
+
+/// Check if the type variable `ptr` occurs in `t`
+fn occurs_check(ptr: usize, state: &State, t: TypeWrapper) -> bool {
```

Comment:
Nitpick: from what I can see it seems like the pattern elsewhere has been to pass the `state` as the first argument. It possibly makes sense to continue that here?